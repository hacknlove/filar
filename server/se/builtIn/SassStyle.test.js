const parser = new (require("linkedom").DOMParser)();
const { ServerElementsMap } = require("../common");
const { processAllElements } = require("../../tree/processAllElements");
const config = require("../../config");
const { join } = require("path");
const { readFile } = require("fs-extra");
const { saveIfDifferent } = require("../../common/saveIfDifferent");
const { globAsync } = require("../../common/globAsync");
const { createFakeContext } = require("../../../test/fakeContext");

jest.mock("fs-extra");
jest.mock("../../common/saveIfDifferent");
jest.mock("../../config", () => ({
  from: __dirname,
  staticDir: ".build/static",
}));
jest.mock("../../common/globAsync");

const SassStyle = require("./SassStyle.se");

ServerElementsMap.set("SassStyle", SassStyle);

describe("processServerElement", () => {
  it("compiles and adds style to head", async () => {
    const document = parser.parseFromString(`
            <html>
                <head></head>
                <body>
                    <SassStyle embedded color1="red" color2="#fabada"/>
                </body>
            </html>s
        `);

    globAsync.mockResolvedValue(["test.scss", "test2.scss"]);

    readFile.mockResolvedValueOnce(`h1 { color: var(--color1); }`);
    readFile.mockResolvedValueOnce(`h2 { color: var(--color2); }`);

    await processAllElements(document, createFakeContext());

    expect(document.querySelector("style")).not.toBeNull();
    expect(document.toString()).toMatchSnapshot();
  });

  it("compiles and adds link to head", async () => {
    const document = parser.parseFromString(`
            <html>
                <head></head>
                <body>
                    <SassStyle color1="red" color2="#fabada"/>
                </body>
            </html>s
        `);

    globAsync.mockResolvedValue(["test.scss", "test2.scss"]);

    readFile.mockResolvedValueOnce(`h1 { color: var(--color1); }`);
    readFile.mockResolvedValueOnce(`h2 { color: var(--color2); }`);

    await processAllElements(document, createFakeContext());

    expect(document.querySelector("style")).toBeNull();
    expect(document.querySelector("link")).not.toBeNull();
    expect(saveIfDifferent).toHaveBeenCalledWith(
      join(
        config.from,
        ".build",
        "static",
        document.querySelector("link").getAttribute("href")
      ),
      expect.any(String)
    );
    expect(document.toString()).toMatchSnapshot();
  });
});
