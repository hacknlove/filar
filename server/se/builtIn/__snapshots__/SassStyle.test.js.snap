// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processServerElement compiles and adds link to head 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
            <html>
                <head><link rel="stylesheet" href="/cb14b40216f1a7303bf59c768e9d616e8b5608db193cfe93324f4e159f4e2661.css" type="text/css" /></head>
                <body>
                    
                </body>
            </html>s
        "
`;

exports[`processServerElement compiles and adds style to head 1`] = `
"<?xml version="1.0" encoding="utf-8"?>
            <html>
                <head><style>:root{--color1: red;--color2: #fabada}h1{color:var(--color1)}h2{color:var(--color2)}</style></head>
                <body>
                    
                </body>
            </html>s
        "
`;
